contracts/oracle/src/contract/feed/leg_cmd.rs

```diff
@@ -0,0 +1,174 @@
+use super::price_querier::PriceQuerier;
+use crate::ContractError;
+use finance::{
+    currency::{AnyVisitorPair, Currency},
+    price::{base::BasePrice, Price},
+};
+use serde::de::DeserializeOwned;
+use swap::SwapGroup;
+
+pub struct LegCmd<OracleBase, Querier>
+where
+    OracleBase: Currency,
+    Querier: PriceQuerier,
+{
+    pub price_querier: Querier,
+    pub stack: Vec<BasePrice<SwapGroup, OracleBase>>,
+}
+
+impl<OracleBase, Querier> AnyVisitorPair for &mut LegCmd<OracleBase, Querier>
+where
+    OracleBase: Currency + DeserializeOwned,
+    Querier: PriceQuerier,
+{
+    type Output = Option<BasePrice<SwapGroup, OracleBase>>;
+    type Error = ContractError;
+
+    fn on<B, Q>(self) -> Result<Self::Output, Self::Error>
+    where
+        B: Currency + DeserializeOwned,
+        Q: Currency + DeserializeOwned,
+    {
+        let branch_price = self
+            .stack
+            .iter()
+            .enumerate()
+            .rev()
+            .find_map(|(i, parent_bprice)| {
+                Price::<Q, OracleBase>::try_from(parent_bprice)
+                    .ok()
+                    .map(|parent_price| {
+                        self.price_querier
+                            .price::<B, Q>()
+                            .map(|res_price| res_price.map(|price| (i + 1, price * parent_price)))
+                    })
+            })
+            .transpose()
+            .map(Option::flatten)?;
+
+        let idx_price = branch_price.map_or_else(
+            || {
+                self.price_querier
+                    .price::<B, OracleBase>()
+                    .map(|res_price| res_price.map(|price| (0, price)))
+            },
+            |idx_price| Ok(Some(idx_price)),
+        )?;
```

Comment:
I am afraid here the logic is a source of bugs. There are three cases for `branch_price` to be None:
- Q is found in the stack but there is no B-to-Q price at the moment
- Q is not found because Q is a descendant to the top of the stack T, for example, on the previous round price(Q, T) was None and Q was not pushed to the stack
- Q is not found because Q == OracleBase and this is the first time it is invoked
The implementation here assumes it is the last option.

Reply:
IMHO, one way to fix it is to consider two main cases - stack empty or not. If empty, add B-to-Q price only if Q == OracleBase, otherwise skip. If not empty, search for Q. If found, calculate the price, and add it if Some. In both other cases, Q is not found or no price, None.